### PR TITLE
Migrate credential-library nomenclature around targets to credential-source

### PR DESCRIPTION
**Pull request by jefferai**
Outside of updating unit tests, this is not yet tested...that's why it's still draft. However, it's generally ready for review, in terms of paradigm.

Some notes:

* At the CLI/API level I chose to mostly duplicate code rather than add switches on behavior. This both makes it easier to check correctness now and easier to remove the library versions later.
* Within the handler's internals things have switched to the abstract type. However, at the moment I didn't need to replace views or add tables in the database; I simply ensure that what's returned from the existing view matches the interface. Once we get to static credentials the DB conversion can take place.
* I need to do some manual testing, which I will do soon.

### CODE REVIEW

File: internal/target/repository_tcp_target.go [hunk 1]
```diff
@@ -105,10 +105,10 @@ func (r *Repository) CreateTcpTarget(ctx context.Context, target *TcpTarget, opt
 			if len(newCredLibs) > 0 {
 				credLibOplogMsgs := make([]*oplog.Message, 0, len(newCredLibs))
 				if err := w.CreateItems(ctx, newCredLibs, db.NewOplogMsgs(&credLibOplogMsgs)); err != nil {
-					return errors.Wrap(err, op, errors.WithMsg("unable to add credential libraries"))
+					return errors.Wrap(err, op, errors.WithMsg("unable to add credential sources"))
 				}
-				if returnedCredLibs, err = fetchLibraries(ctx, read, t.PublicId); err != nil {
-					return errors.Wrap(err, op, errors.WithMsg("unable to read host sets"))
+				if returnedCredSources, err = fetchCredentialSources(ctx, read, t.PublicId); err != nil {
+					return errors.Wrap(err, op, errors.WithMsg("unable to read credential ources"))
```
Comment (jimlambrt) [inline internal/target/repository_tcp_target.go:111]:
```suggestion
					return errors.Wrap(err, op, errors.WithMsg("unable to read credential sources"))
```

File: internal/target/repository_credential_source_test.go [hunk 2]
```diff
@@ -335,30 +335,30 @@ func TestRepository_DeleteTargetCredentialLibraries(t *testing.T) {
 		lib3 := libs[2]
 
 		projTarget := TestTcpTarget(t, conn, proj.PublicId, "add-existing")
-		_, _, gotCredLibs, err := repo.AddTargetCredentialLibraries(context.Background(), projTarget.PublicId, 1, []string{lib1.PublicId, lib2.PublicId})
+		_, _, gotCredSources, err := repo.AddTargetCredentialSources(context.Background(), projTarget.PublicId, 1, []string{lib1.PublicId, lib2.PublicId})
 		require.NoError(err)
-		assert.Len(gotCredLibs, 2)
+		assert.Len(gotCredSources, 2)
 
 		// Deleting an unassociated library should return an error
```
Comment (jimlambrt) [inline internal/target/repository_credential_source_test.go:342]:
```suggestion
		// Deleting an unassociated source should return an error
```

File: internal/target/repository_credential_source_test.go [hunk 3]
```diff
@@ -335,30 +335,30 @@ func TestRepository_DeleteTargetCredentialLibraries(t *testing.T) {
...
-		delCount, err := repo.DeleteTargetCredentialLibraries(context.Background(), projTarget.PublicId, 2, []string{lib3.PublicId})
+		delCount, err := repo.DeleteTargetCredentialSources(context.Background(), projTarget.PublicId, 2, []string{lib3.PublicId})
 		require.Error(err)
 		assert.True(errors.Match(errors.T(errors.MultipleRecords), err))
 		assert.Equal(0, delCount)
 
 		// Deleting libraries which includes an unassociated libary should return an error
```
Comment (jimlambrt) [inline internal/target/repository_credential_source_test.go:348]:
```suggestion
		// Deleting sources which includes an unassociated source should return an error
```

File: internal/target/repository.go [hunk 4]
```diff
@@ -56,7 +56,7 @@ func NewRepository(r db.Reader, w db.Writer, kms *kms.Kms, opt ...Option) (*Repo
 // with its host set ids and credential library ids.  If the target is not found,
```
Comment (jimlambrt) [inline internal/target/repository.go:56]:
```suggestion
// with its host set ids and credential source ids.  If the target is not found,
```

File: internal/proto/local/controller/api/services/v1/target_service.proto [hunk 5]
```diff
@@ -199,10 +201,64 @@ service TargetService {
...
+    option deprecated = true;
   }
 
+  // AddTargetCredentialSources adds Credential Sources to this Target.
+  // The provided request must include the Target ID to which the Credential
+  // Sources will be added. All Credential Sources added to the provided
+  // Target must be a child of a Store that is in the same scope as this
+  // Target. If the scope or Target IDs are missing, malformed, or reference
+  // non-existing resources, an error is returned. An error is returned if a
+  // Credential Library is attempted to be added to a target that is already
```
Comment (jimlambrt) [inline internal/proto/local/controller/api/services/v1/target_service.proto:215]:
```suggestion
  // Credential Source is attempted to be added to a target that is already
```

File: internal/proto/local/controller/api/services/v1/target_service.proto [hunk 6]
```diff
@@ -199,10 +201,64 @@ service TargetService {
...
+      summary: "Sets the Credential Sources on the Target."
+    };
+  }
+
+  // RemoveTargetCredentialSources removes the Credential Sources from the
+  // specified Target. The provided request must include the Target ID for the
+  // Target from which the Credential Sources will be removed. If the ID is
+  // missing, or malformed, an error is returned.  An error is returned if a
+  // Credential Library is attempted to be removed from the Target when the
+  // Target does not have the Credential Library.
```
Comment (jimlambrt) [inline internal/proto/local/controller/api/services/v1/target_service.proto:251]:
```suggestion
  // Target does not have the Credential Source.
```

File: internal/servers/controller/handlers/targets/target_service.go [hunk 7]
```diff
@@ -565,6 +568,120 @@ func (s Service) RemoveTargetCredentialLibraries(ctx context.Context, req *pbs.R
 	return &pbs.RemoveTargetCredentialLibrariesResponse{Item: item}, nil
 }
 
+// AddTargetCredentialSources implements the interface pbs.TargetServiceServer.
```
Comment (jimlambrt) [inline internal/servers/controller/handlers/targets/target_service.go:571]: Just a reminder to add unit tests for these new function.

**Review by jimlambrt — CHANGES_REQUESTED** [review]
Really amazing work and ty!  Just a few minor bits.

File: internal/target/credential_source.go [hunk 8]
```diff
@@ -0,0 +1,12 @@
+package target
+
+import "github.com/hashicorp/boundary/internal/credential"
+
+// CredentialSource is an interface that can be implemented by both a library
+// and a singular credential.
```
[the remark this answers is not in the document]
Comment (jefferai) [reply internal/target/credential_source.go:6]: Maybe, but let's leave that for a follow-on as it's also in a PR state.

File: internal/servers/controller/handlers/targets/target_service.go (hunk 7 above)
Comment (jefferai) [reply internal/servers/controller/handlers/targets/target_service.go:571]: Done

**Review by jimlambrt — DISMISSED** [review]

File: internal/cmd/commands/targetscmd/funcs.go [hunk 9]
```diff
@@ -65,12 +69,24 @@ func extraSynopsisFuncImpl(c *Command) string {
...
 		case strings.HasPrefix(c.Func, "remove"):
 			in = "Remove credential libraries from"
 		}
+		return wordwrap.WrapString(fmt.Sprintf("%s a target. DEPRECATED; use the -sources version instead.", in), base.TermWidth)
+
+	case "add-credential-sources", "set-credential-sources", "remove-credential-sources":
+		var in string
+		switch {
+		case strings.HasPrefix(c.Func, "add"):
+			in = "Add credential sourcse to"
```
Comment (mgaffney) [inline internal/cmd/commands/targetscmd/funcs.go:84]:
```suggestion
			in = "Add credential sources to"
```

File: internal/proto/local/controller/api/services/v1/target_service.proto [hunk 10]
```diff
@@ -199,10 +201,64 @@ service TargetService {
...
+    option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {
+      summary: "Sets the Credential Sources on the Target."
+    };
+  }
+
+  // RemoveTargetCredentialSources removes the Credential Sources from the
+  // specified Target. The provided request must include the Target ID for the
+  // Target from which the Credential Sources will be removed. If the ID is
+  // missing, or malformed, an error is returned.  An error is returned if a
+  // Credential Library is attempted to be removed from the Target when the
```
Comment (mgaffney) [inline internal/proto/local/controller/api/services/v1/target_service.proto:250]:
```suggestion
  // Credential Source is attempted to be removed from the Target when the
```

File: internal/servers/controller/handlers/targets/target_service.go [hunk 11]
```diff
@@ -1277,22 +1394,42 @@ func toProto(ctx context.Context, in target.Target, m []*target.TargetSet, l []*
...
-					Id:                cl.GetCredentialLibraryId(),
-					CredentialStoreId: cl.StoreId,
+					Id:                cs.Id(),
+					CredentialStoreId: cs.CredentialStoreId(),
 				})
 			case credential.IngressPurpose, credential.EgressPurpose:
 				// TODO: When we support other purposes add them to different fields here.
 			default:
-				return nil, errors.New(errors.Internal, op, fmt.Sprintf("unrecognized purpose %q for credential library on target", cl.GetCredentialPurpose()))
+				return nil, errors.New(errors.Internal, op, fmt.Sprintf("unrecognized purpose %q for credential library on target", cs.CredentialPurpose()))
```
Comment (mgaffney) [inline internal/servers/controller/handlers/targets/target_service.go:1417]:
```suggestion
				return nil, errors.New(errors.Internal, op, fmt.Sprintf("unrecognized purpose %q for credential source on target", cs.CredentialPurpose()))
```

File: internal/target/credential_library.go [hunk 12]
```diff
@@ -85,3 +89,23 @@ type TargetLibrary struct {
...
+func (ts *TargetLibrary) Id() string {
+	return ts.CredentialLibraryId
+}
+
+// CredentialStoreId returns the ID of the store containing the library
+func (ts *TargetLibrary) CredentialStoreId() string {
+	return ts.StoreId
+}
+
+// CredentialPurpose returns the purpose of the credentaial
```
Comment (mgaffney) [inline internal/target/credential_library.go:103]:
```suggestion
// CredentialPurpose returns the purpose of the credential
```

**Review by mgaffney — DISMISSED** [review]
Looks good. Just some minor fixes to comments.

**Pull request merged**